nomad/resource_job.go

```diff
@@ -258,109 +265,129 @@ func resourceJobRegister(d *schema.ResourceData, meta interface{}) error {
 	d.Set("namespace", job.Namespace)
 	d.Set("modify_index", strconv.FormatUint(resp.JobModifyIndex, 10))
 
-	evalId := resp.EvalID
-	if d.Get("detach").(bool) {
-		d.Set("deployment_id", "")
-		d.Set("deployment_status", "")
-	} else {
-		log.Printf("[DEBUG] will monitor deployment of job '%s'", *job.ID)
-		stateConf := &resource.StateChangeConf{
-			Pending:    []string{"monitoring_deployment", "monitoring_evaluation"},
-			Target:     []string{"job_scheduled_without_deployment", "deployment_successful"},
-			Refresh:    deploymentStateRefreshFunc(client, evalId),
-			Timeout:    timeout,
-			Delay:      10 * time.Second,
-			MinTimeout: 3 * time.Second,
-		}
-
-		state, err := stateConf.WaitForState()
+	if d.Get("detach") == false && resp.EvalID != "" {
+		log.Printf("[DEBUG] will monitor scheduling/deployment of job '%s'", *job.ID)
+		deployment, err := monitorDeployment(client, timeout, resp.EvalID)
 		if err != nil {
 			return fmt.Errorf(
-				"error waiting for job '%s' to deploy successfully: %s",
+				"error waiting for job '%s' to schedule/deploy successfully: %s",
 				*job.ID, err)
 		}
-
-		deployment := state.(*api.Deployment)
-
 		if deployment != nil {
 			d.Set("deployment_id", deployment.ID)
 			d.Set("deployment_status", deployment.Status)
 		} else {
-			d.Set("deployment_id", "")
-			d.Set("deployment_status", "")
+			d.Set("deployment_id", nil)
+			d.Set("deployment_status", nil)
 		}
 	}
 
 	return resourceJobRead(d, meta) // populate other computed attributes
 }
 
-// deploymentStateRefreshFunc returns a resource.StateRefreshFunc that is used to watch
-// the deployment from a job create/update
-func deploymentStateRefreshFunc(client *api.Client, initialEvalId string) resource.StateRefreshFunc {
+// monitorDeployment monitors the evalution(s) from a job create/update and,
+// if they result in a deployment, monitors that deployment until completion.
+func monitorDeployment(client *api.Client, timeout time.Duration, initialEvalId string) (*api.Deployment, error) {
```

Comment:
My editor is complaining that `initialEvalId` should be `initialEvalID`
```suggestion
func monitorDeployment(client *api.Client, timeout time.Duration, initialEvalID string) (*api.Deployment, error) {
```